lib/phoenix/endpoint.ex

```diff
@@ -475,6 +476,12 @@ defmodule Phoenix.Endpoint do
           ]
       end
 
+      plug :match
+      plug :dispatch_if_matched
+
+      def dispatch_if_matched(%Plug.Conn{} = conn, opts),
+        do: if(Map.has_key?(conn.private, :plug_route), do: dispatch(conn, opts), else: conn)
```

Comment:
ibid.

Reply:
I am almost sure we are using a private API here too.

Perhaps it is best to ditch `Plug.Router` and build a mini router, which we already have in `__handler__`, except we will call `:socket_dispatch` here. The difference is that `socket_dispatch`  invokes the underlying plugs instead of returning metadata. 

Then all we need to do is to add a `plug :socket_dispatch` here.

Reply:
You're correct that this is also part of Plug.Router's private API, though it's part of the same problem that we have with the private `defp do_match` called out below.

The main reason I wanted to appeal to Plug.Router here is to get the benefit of its excellent `/path/:id` parameter parsing, which we'd need to replicate. 

I would think that adding in a `use Plug.Router, continue_on_no_match: true` style option would solve this quite handily, and also make Plug.Router useful within plug pipelines in a general case (this is something I've come up against before as well, in other 'pure Plug' projects).

Reply:
Is there a reason that you're opposed to adding that to Plug.Router? To be clear, I'm happy to do the work for that (it's already done other than tests locally).

Reply:
1. We already have the match logic implemented today
2. We won't use private APIs
3. We won't rock the boat too much which means...
4. We will get this merged sooner

:)

Reply:
In fact, I am even thinking we may even move the whole `socket` stuff to the router, so not rocking the boat too much sounds even better if the goal is to deprecate this eventually.